### PR TITLE
DuckDuckGo Plugin

### PR DESCRIPTION
**Pull request by paddatrapper**
Created a DuckDuckGo Instant Answer query plugin. Also fixed the DB connection to work with newer versions of sqlalchemy.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -6,3 +6,8 @@
 /local.ini
 /logs/
 dropin.cache
+/pyenv/
+/build/
+/Ibid.egg-info/
+/dist/
+/bothome/
```
Comment (stefanor) [inline .gitignore]: sort?

File: ibid/db/types.py [hunk 2]
```diff
@@ -8,28 +8,28 @@
 def monkey_patch():
     import sqlalchemy.dialects.postgresql
     sqlalchemy.dialects.postgresql.dialect.ischema_names['citext'] = IbidUnicodeText
-    def postgres_visit_IBID_VARCHAR(self, type_):
+    def postgres_visit_IBID_VARCHAR(self, type_, **kw):
         if type_.case_insensitive:
             return 'CITEXT'
         else:
             return self.visit_VARCHAR(type_)
```
Comment (stefanor) [inline ibid/db/types.py]: Shouldn't we be passing through the `**kw` here?

File: ibid/plugins/ddg.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
+# Copyright (c) 2008-2011, Michael Gorven, Stefano Rivera
```
Comment (stefanor) [inline ibid/plugins/ddg.py]: Take some ownership :) This should be 2016, paddatrapper

File: ibid/plugins/ddg.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+from ibid.plugins import Processor, match
+from ibid.utils import decode_htmlentities, json_webservice
+
+features = {'duckduckgo': {
+    'description': u'Retrieves results from DuckDuckGo',
+    'categories': ('lookup', 'web', 'calculate', ),
+}}
+
+default_user_agent = 'Mozilla/5.0'
+default_referer = "http://ibid.omnia.za.net/"
```
Comment (stefanor) [inline ibid/plugins/ddg.py]: Stick to one quote style for the whole file. Single quotes tend to be preferred.

File: ibid/plugins/ddg.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+            'q': query,
+            't': 'ibid',
+            'rsz': resultsize,
+            'format': 'json',
+        }
+        if country is not None:
+            params['gl'] = country
+
+        headers = {'referer': self.referer}
+        return json_webservice('http://api.duckduckgo.com', params, headers)
```
Comment (stefanor) [inline ibid/plugins/ddg.py]: No https? :(

File: ibid/plugins/ddg.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+    @match(r'^(ddg|duckduckgo(?:\.com?))?(?:\.([a-z]{2}))?\s+(?:for\s+)?(.+?)$')
```
Comment (stefanor) [inline ibid/plugins/ddg.py]: err the ddg | duckduckgo bit is optional?

File: ibid/plugins/ddg.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+            items = self._ddg_api_search(query, country=country)
+        except BadStatusLine:
+            event.addresponse(u"DuckDuckGo appears to be broken (or more likely, my connection to it)")
+            return
+
+        results = []
+        topic = "Results"
+        for item in items[topic]:
+            title = item['Text']
+            results.append(u'"%s" %s' % (decode_htmlentities(title), item['FirstURL']))
```
Comment (stefanor) [inline ibid/plugins/ddg.py]: The API response is HTML encoded? o_O

File: ibid/plugins/ddg.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        topic = "RelatedTopics"
+        for i in range(5):
+            if i >= len(items[topic]):
+                break
```
Comment (stefanor) [inline ibid/plugins/ddg.py]: `for i in range(max(5, len(items[topic])))` ?

File: ibid/plugins/ddg.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+            try:
+                title = items[topic][i]['Text']
+                results.append(u'%s' % (decode_htmlentities(title)))
```
Comment (stefanor) [inline ibid/plugins/ddg.py]: The string substitution doesn't seem to be doing anything useful

File: ibid/plugins/ddg.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+            except KeyError:
+                pass
```
Comment (stefanor) [inline ibid/plugins/ddg.py]: Why wolud that happen?

File: ibid/plugins/ddg.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+            'v': '1.0',
+            'q': query,
+            't': 'ibid',
+            'rsz': resultsize,
+            'format': 'json',
+        }
+        if country is not None:
+            params['gl'] = country
+
+        headers = {'referer': self.referer}
```
Comment (stefanor) [inline ibid/plugins/ddg.py]: Do they actually require a referer header?

File: .gitignore (hunk 1 above)
Comment (paddatrapper) [inline .gitignore]: As in put them in alphabetical order?

**Comment by paddatrapper** [comment]
All issues corrected

**Pull request merged**